avr-hal-generic/src/serial.rs

```diff
@@ -1,5 +1,75 @@
 //! Serial Implementations
 
+use core::cmp::Ordering;
+
+// Clock is needed because the calculations needs to take core clock into account
+#[derive(Debug, Clone, Copy)]
+pub struct Baudrate<CLOCK> {
+    pub ubrr: u16,
+    pub u2x: bool,
+    pub _clock: ::core::marker::PhantomData<CLOCK>,
+}
+
+impl<CLOCK: crate::clock::Clock> PartialEq for Baudrate<CLOCK> {
+    fn eq(&self, other: &Self) -> bool {
+        self.value() == other.value()
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Eq for Baudrate<CLOCK> {}
+
+impl<CLOCK: crate::clock::Clock> PartialOrd for Baudrate<CLOCK> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.value().cmp(&other.value()))
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Ord for Baudrate<CLOCK> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.value().cmp(&other.value())
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Baudrate<CLOCK> {
+    pub fn new(baud: u32) -> Baudrate<CLOCK> {
+        // Try u2x mode first
+        let mut baud_setting = (CLOCK::FREQ / 4 / baud - 1) / 2;
+        let mut u2x = true;
+
+        // Switch back from u2x if required
+        if ((CLOCK::FREQ == 16_000_000) && (baud == 57_600)) || (baud_setting > 4095) {
+            u2x = false;
+            baud_setting = (CLOCK::FREQ / 8 / baud - 1) / 2;
+        }
```

Comment:
This is what Arduino is doing but especially that workaround for their bootloader seems suspicious to me.  We have to remember that not everyone using the HAL is using it for Arduino boards so we should be careful about shipping Arduino workarounds everywhere by default.  I'll look deeper into this and share my results.

Reply:
In general, enabling U2X consistently yields clock rates with equal or less error than with disabled U2X so I'd say it should be the default where possible.  I.e. it makes sense to ship this algorithm in the `Baudrate::new()` method.

The workaround (`(CLOCK::FREQ == 16_000_000) && (baud == 57_600))`) is really not something we want to include by default, however.  It actually makes the clock error worse, going from `-0.8%` (with u2x) to `+2.1%` (without u2x).  This is (apparently) necessary because the USB-to-Serial bridge on some arduino boards has a firmware with a similar clock error and the `-0.8%` in the other direction would make the total error `3%` which is too much, I guess.

My suggestion would be to add a custom `BaudrateArduinoExt` trait where the fix is applied.  On the affected boards, `BaudrateArduinoExt` is pulled into the prelude instead of `BaudrateExt`.

```rust
pub trait BaudrateArduinoExt {
    fn into_baudrate<CLOCK: crate::clock::Clock>(self) -> Baudrate<CLOCK>;
}

impl BaudrateArduinoExt for u32 {
    fn into_baudrate<CLOCK: crate::clock::Clock>(self) -> Baudrate<CLOCK> {
        let br = Baudrate::new(self);

        // hardcoded exception for 57600 for compatibility with the bootloader
        // shipped with the Duemilanove and previous boards and the firmware
        // on the 8U2 on the Uno and Mega 2560.
        //
        // https://github.com/arduino/ArduinoCore-avr/blob/3055c1efa3c6980c864f661e6c8cc5d5ac773af4/cores/arduino/HardwareSerial.cpp#L123-L132
        if CLOCK::FREQ == 16_000_000 && br.ubrr == 34 && br.u2x {
            // (CLOCK::FREQ / 8 / 57600 - 1) / 2 == 16
            Baudrate::with_exact(false, 16)
        } else {
            br
        }
    }
}
```

The only problem I see with this is that it's a bit "magical" and makes it quite non-obvious which version is used.  Though the alternative of requiring board users to explicitly do something different depending on the hardware they are using, is against the idea of having a HAL in the first place ...

Reply:
I done this. Now I'm testing `leonardo-serial` example. After upload, arduino is not recognized until I will reset. After that I need to cleanup board to works fine without reset.

I'm not sure if it's regression, serial example (need to spend some more time on example itself) or problem with `.elf`

Reply:
Right, you're probably trying to access the serial console directly via the USB cable which you use to flash code to the board?  That's not going to work on Leonardo as it is different from the other Arduino boards:  On other Arduinos there is a separate USB-to-Serial converter between the USB port and the actual MCU.  On Leonardo, the USB port is directly connected to the MCU (ATmega32U4) and for Arduino C code, the MCU directly presents itself to the host as a USB-to-Serial converter.  In Rust we don't do that (at this time at least).  Instead, what you're working with right now is the real USART peripheral connected to the pins D0 (RX) and D1 (TX).  So to access it you'll need a separate USB-to-Serial converter to connect to those pins.

Reply:
Nice, I have some usb to ttl form my old ESP project. After quick test I was able to send some chars with `let baudrate = 9600_u32.into_baudrate();` and receive correct response: Will test below combination - with Ext. impl and with exact ubrr
baund| UBRR|tested|
-- | -- |--|
300 | 767 | no|
600 | 383 | no|
1200 | 191 |no|
2400 | 95 | no|
4800 | 47 | no|
9600 | 23 | x |
14400 | 15 | no|
19200 | 11 | no|
28800 | 7 | no|
38400 | 5 | no|
57600 | 3 | no|
76800 | 2 | no|
115200 | 1|no|

Reply:
For now I have three failing baunds `14400`, `28800` and `76800`

Reply:
You might want to check whether the chip on your USB-to-Serial converter properly supports those rates before diving into the AVR side.  Those three don't look like common values to me, so maybe there is some problem ...

Reply:
That aside, we should also run a test on an Arduino Uno with the special implementation to check whether that's working as intended.  If you don't have a device on hand, I can take care of that :)

Reply:
I will be able to test with uno in the end of this week. So if with your help we can be faster.

PL-2303HX should support some of the bauds. http://www.prolific.com.tw/UserFiles/files/ds_pl2303HXD_v1_4_4.pdf
Not sure about linux drivers. Right now I'm on 5.4.0-52-generic and I can see it. Not sure if supports all bauds.

Reply:
Linux will just pass the baudrate to the CDC-ACM endpoint exposed by your PL-2303HX.  What the device makes of it is the important part, some just select the closest baudrate, others properly configure their clock for the exact selected speed.

From the datasheet you linked, at least 14400 and 28800 should work though ... You don't by any chance have an oscilloscope at hand to check the signal?  I'm a bit short on time this and the next week.  Though after that I could run some intrusive tests to see.

That said, I wouldn't block this PR's progress on these few baudrates though.  As long as all common rates are well supported, we are good I'd say.

Reply:
I do not have now any oscilloscope now. With time I will buy some. Now I don't want to invest to much now.

Reply:
Let's open an issue for this an resolve it at a later time: #99